Merge branch DeviceSettingsApiPUT into dev

sebastienvolont opened the pull request:
Merge branch DeviceSettingsApiPUT into dev. This adds the features of adding and modifying a device and some global correctings errors / warnings

sebastienvolont committed:
Adding API use with fetch to display devices list from database

The API is used with fetch and react in order to display devices on devicelist page. The get request will be used thereafter to store and display informations in forms from DeviceSettings component

sebastienvolont committed:
(frontend) Api used to edit a device configuration with PUT method. Page refresh is missing

This feature permit the user to edit his devices with a configuration menu.
The page refresh is missing in the feature and will be added as soon as possible

sebastienvolont committed:
(frontend) Add feature "add device" and separating edit and add features in two distinct components, page refresh is missing

sebastienvolont committed:
(frontend) Fix React key warning in console

The issue came from that we used an image component, and a key was needed to differentiate itself from other components created with the map function.

sebastienvolont committed:
(frontend) Update AddDevice to refresh the list devices display

The class method « displayDevice » is passed in the component props « AddDevice » to refresh display once a device is added.

sebastienvolont committed:
(frontend) Update the device configuration feature, list of devices is refreshing and visual display of the form is correct

This update is in order to fix the display of the list when a device is updated.
It misses form fields verification and props are not yet updating in form fields configuration.

sebastienvolont committed:
(frontend) Fix fields not updating and some fields data checks

This commit solves the issue that the fields weren't up to date after a change.
In addition, it adds some fields checks the validity of the ipv4 and the name of the device.

sebastienvolont committed:
(frontend) Changing DeviceSettings and AddDevice into a single component named « DevicesManagement ».

This commit is done in order to reduce code redundancy with add and modify device components into a single component named « DevicesManagement ».

sebastienvolont committed:
(fix) Using « Link » tag in place of « a » tag on Details page in order to not have to refreshpage

mwath commented on web/src/components/DevicesManagement.tsx:
```diff
@@ -0,0 +1,122 @@
+import {Modal, Button, Form, Image, Alert} from 'react-bootstrap'
+import { RiSettings2Line } from "react-icons/ri";
+
+class DevicesManagement extends React.Component {
```
Il manque les types des props/state sur le component.

mwath replied:
`React.Component<{}, DevicesManagementInterface>`

mwath commented on web/src/components/DevicesManagement.tsx:
```diff
@@ -0,0 +1,122 @@
+      setValidAlertShow: false,
+      messageErreur: "",
+        }
+    };
```
C'est possible de faire ça sans constructeur:
```Typescript
state: TypeState = {
      modalShow: false,
      setModalShow: false,
      validAlertShow: false,
      invalidAlertShow: false,
      setInvalidAlertShow: false,
      setValidAlertShow: false,
      messageErreur: "",
}
```

mwath commented on web/src/components/DevicesManagement.tsx:
```diff
@@ -0,0 +1,122 @@
+    handleValidAlertShow() {
+      if (this.invalidAlertShow) {
+        this.setState({validAlertShow: false});
+    } else {
```
indentation ?

mwath replied:
Il faudrait voir pour formater le fichier 🤔

mwath commented on web/src/components/DevicesManagement.tsx:
```diff
@@ -0,0 +1,122 @@
+        }
+
+        else {
+          fetch("https://localhost/api/devices" + (isNaN(id) ? "" : `/${id}`), {method: methodParameter,
```
Il ne faut pas utiliser localhost.
A la place, utilise le chemin absolu: `/api/devices`

mwath commented on web/src/components/FormDeviceOptions.tsx:
```diff
@@ -1,57 +1,67 @@
-                                Définir le nom de l'appareil.
-                            </Form.Text>
-                        </Form.Group>
+class FormDeviceOptions extends React.Component {
+    constructor(props) {
+        super(props)
+        this.state = {
+            nameDevice : props.name,
+            ipDevice : props.ipAddress,
+            idDevice : props.id,
+            groupId : props.groupId,
+            type: props.type,
+        }
+    }
```
Pareil ici, les types et le constructeur n'est pas nécessaire.

mwath commented on web/src/pages/Details.tsx:
```diff
@@ -37,13 +42,8 @@ class Details extends React.Component {
         console.log(devices);
-        let id = this.props.match.params.id
-        console.log(id);
         console.log(typeof (id));
```
S'ils ne sont plus nécessaire, il faudrait enlever les console.log

mwath commented on web/src/pages/ListDevices.tsx:
```diff
@@ -63,9 +68,9 @@ class ListDevices extends React.Component<{}, ListDevicesInterface> {
-
+	render() {
+		let {devices} = this.state;
+		console.log(devices);
```
logs

mwath commented on web/src/pages/ListDevices.tsx:
```diff
@@ -86,14 +91,13 @@ class ListDevices extends React.Component<{}, ListDevicesInterface> {
                     {this.state.devicesShown.map((device: Device, i: number) => (
-                        <DeviceBox img={[<BsLamp/>]} name={device.name} key={i} state="success" ipAddress={device.ip}/>
-                    ))}
+                        <DeviceBox key={device.id} img={[<BsLamp key={device.id}/>]} name={device.name} state="success" ipAddress={device.ip} groupId={device.groupId} type={device.type} id={device.id} displayDevice={this.displayDevice.bind(this)}/>
```
Pourquoi tu mets la propriété `key` à la fois sur l'élément `DeviceBox` et sur l'image `BsLamp` ?

sebastienvolont replied:
> Pourquoi tu mets la propriété `key` à la fois sur l'élément `DeviceBox` et sur l'image `BsLamp` ?

J'ai rajouté la propriété ``key`` à l'image ``BsLamp`` qui est un composant react car il me semble que ce composant a besoin d'une clé (``key``) lors de l'utilisation de la fonction map() pour permettre de différencier les différents composants montés d'image ``BsLamp``  pour chaque appareil contenus dans la table "devices" de la base de données.

Sans cette clé, le composant provoquait un « warning » dans la console de développement Chrome.

mwath reviewed (changes requested):
Il y a juste 2-3 trucs à changer et puis c'est bon.
N'oublie pas de formater les fichiers modifiés !

sebastienvolont committed:
(fix) Required changes for Pull request review #15

This commit contains correct 2 spaces indentation, types precision, removal of constructors and console.log, absolute path api.

sebastienvolont committed:
(fix) Adding changes from Pull Request Review  #15

This commit adds the use of interface for state fields in class components.

sebastienvolont committed:
(fix) fix changes Pull Request Review #15

This commit fixes import file without extensions and adds types in class FormDeviceOptions component.

mwath reviewed (approved):
Parfait
